### DISCUSSION TITLE
Globally replace IRI with a new value

### DISCUSSION THREAD

**Discussion by reckart** [eclipse-rdf4j/rdf4j#4480] [Q&A] [answered]
Sometimes, it may be necessary to change the value of an IRI, e.g. because there is a typo in it or because an initially automatically generated IRI should be replaced with something more semantic.

What would be the best and most efficient way of using the RDF4J API to globally replace all occurrences of a given IRI with a new value?

**Comment by kenwenzel** [comment]
I bet that a SPARQL update would be the fastest.

**Reply by hmottestad** [reply]
Using IsolationLevels.NONE will also help with performance as long as you are ok with the lack of isolation and potential issues with rollback.

**Reply by abrokenjester** [reply]
Agreed, something like:

```sparql
DELETE { <oldIRI> ?p ?o }
INSERT { <newIRI> ?p ?o }
WHERE { <oldIRI> ?p ?o }
```

Rinse, repeat with similar for the occurrences in predicate and object positions (you could _probably_ do it in a single SPARQL update but with readability in mind I'd advise to just go with three separate updates, and just combine them in a single transaction).

**Comment by abrokenjester** [comment, marked as answer]
Alternatively, directly using the API:

```java
conn.begin();
conn.getStatements(oldIRI, null, null).forEach(st -> { conn.remove(st); conn.add(newIRI, st.getPredicate(), st.getObject()); });
conn.getStatements(null, oldIRI, null).forEach(st -> { conn.remove(st); conn.add(st.getSubject(), newIRI, st.getObject()); });
conn.getStatements(null, null, oldIRI).forEach(st -> { conn.remove(st); conn.add(st.getSubject(), st.getPredicate(), newIRI); });
conn.commit();
```

**Reply by reckart** [reply]
I think I like actual code the best :) Thank you!
